bdai_ros2_wrappers/bdai_ros2_wrappers/subscription.py

```diff
@@ -78,3 +81,137 @@ def wait_for_message(
+
+    Args:
+        node (Node): the node being attached
+        topics (list): List of topics
+        mtypes (list): List of message types, one for each topic.
+        kwargs: additional arguments, including
+            delay  (float) The delay in seconds for which the messages
+                could be synchronized.
+            allow_headerless (bool): Whether it's ok for there to be
+                no header in the messages.
+            sleep (float) the amount of time to wait before checking
+                whether messages are received
+            timeout (float or None): Time in seconds to wait. None if forever.
+                If exceeded timeout, self.messages will contain None for
+                each topic.
+            latched_topics (set): a set of topics for which the publisher latches (i.e.
+                sets QoS durability to transient_local).
+
+    """
+    return _WaitForMessages(node, topics, mtypes, **kwargs).messages
+
+
+class _WaitForMessages:
+    def __init__(
+        self,
+        node: rclpy.node.Node,
+        topics: typing.List,
+        mtypes: typing.List,
+        queue_size: int = 10,
+        delay: float = 0.2,
+        allow_headerless: bool = False,
+        sleep: float = 0.5,
+        timeout: typing.Optional[float] = None,
+        verbose: bool = False,
+        exception_on_timeout: bool = False,
+        latched_topics: typing.Optional[typing.Set] = None,
+        callback_group: typing.Optional[rclpy.callback_groups.CallbackGroup] = None,
+    ) -> None:
+        self.node = node
+        self.messages: typing.Optional[typing.Tuple] = None
+        self.verbose = verbose
+        self.topics = topics
+        self.timeout = timeout
+        self.exception_on_timeout = exception_on_timeout
+        self.has_timed_out = False
+        if latched_topics is None:
+            latched_topics = set()
+        self.latched_topics = latched_topics
+        self.logger = rclpy.impl.rcutils_logger.RcutilsLogger(name="wait_for_messages")
+
+        if self.verbose:
+            self.logger.info("initializing message filter ApproximateTimeSynchronizer")
+        self.subs: typing.Optional[typing.List] = [
+            self._message_filters_subscriber(mtype, topic, callback_group=callback_group)
+            for topic, mtype in zip(topics, mtypes, strict=True)  # type: ignore
+        ]
+        self.ts = message_filters.ApproximateTimeSynchronizer(
+            self.subs,
+            queue_size,
+            delay,
+            allow_headerless=allow_headerless,
+        )
+        self.ts.registerCallback(self._cb)
+
+        try:
+            self._start_time = self.node.get_clock().now()
+            rate = self.node.create_rate(1.0 / sleep)
+            while self.messages is None and not self.has_timed_out:
+                if self.check_messages_received():
+                    break
+                rate.sleep()
+        finally:
+            self.destroy_subs()
```

Comment:
@kaiyu-zheng meta: by turning this into a `future`, resolved by `_cb`, you can decouple and defer much of the wait logic (incl. timeouts and exceptions) to the user. No need to do expensive polling either. See `wait_for_message_async` for reference on how a future can be instantiated and resolved. Also on how to tie subscription cleanup to future resolution.

Reply:
By user, do you mean the user who called "wait_for_messages"? Why would the user want to handle this logic? My take is from the user's perspective, they just want to wait till messages are returned. 

Do you think making this change would affect how the test is written? I personally like the way `wait_for_messages` is used from the user perspective as shown in the test.

Reply:
This is great advice. It simplified the code so much, and doesn't really affect the user interface. 

Refactored. Thanks!!